feat: add paints support

LosFarmosCTL opened the pull request:
# Description

This is a continuation building upon #86, since the original author is not working on that PR anymore:

<details>
<summary>Discord screenshots</summary>

![image](https://user-images.githubusercontent.com/80157503/168497794-5a2fdc82-63fa-43f8-9c38-aa23b82ee358.png)
![image](https://user-images.githubusercontent.com/80157503/168497810-20ca0d2f-988e-43e8-82f3-5749f6b7484f.png)
</details>

## Checklist

- [x] fetch paints from the API
- [x] linear-gradient paints
- - [x] translate css angle to qt's 2-point style gradients
- - [x] figure out how exactly css deals with repeating gradients and their color stops, to translate that to qt as well
- [x] radial-gradient paints
- [x] drop shadows (i *might* have a way to get these to work, its only theoretical though and i have not confirmed if it works 100%)
- [x] url paints
- - [x] with animations (maybe)
- [x] setting to toggle paints off
- [x] separate the colon from the username itself and only draw paint on the nametag itself to match webchat behavior
- [x] some refactors xd

williambulin committed:
feat: Merge branch 'feature/paints' into feat/paints

LosFarmosCTL committed:
feat(paints): add class framework to store paints

LosFarmosCTL committed:
refactor(api): move SeventvBadges to SeventvCosmetics to include paints

LosFarmosCTL committed:
refactor(api): clean up SeventvCosmetics

LosFarmosCTL committed:
feat(api): add function to calculate rgba color from decimal

LosFarmosCTL committed:
feat(paints): implement linear-gradient paint

LosFarmosCTL committed:
fix(nametag): remove random example paints

LosFarmosCTL committed:
fix(paints): add LinearGradientPaint constructor

LosFarmosCTL committed:
fix(paints): add prototype radial gradient and url paints implementation

LosFarmosCTL committed:
feat(api): load paints from the 7tv API

LosFarmosCTL committed:
fix(api): remove cv-paste error

LosFarmosCTL committed:
fix(api): store pointers to paints

LosFarmosCTL committed:
refactor(paints): mark asBrush as override and const

LosFarmosCTL committed:
chore(nametag): clean up old imports

LosFarmosCTL committed:
chore(formatting): run clang-format

AnatoleAM commented:
> oh and no idea how i got zneix commit merged in there, absolutely no Clue about how I messed that up myself 

Seems that you forked the original Chatterino2, rather than this repo. This may or may not make it difficult to merge this pull request in the future. Might be best to fork the correct repo and open a new PR to avoid complications / merging dankness

LosFarmosCTL commented:
> Seems that you forked the original Chatterino2, rather than this repo. This may or may not make it difficult to merge this pull request in the future. Might be best to fork the correct repo and open a new PR to avoid complications / merging dankness

I have cloned the c7 repo, it might be some dankness around the github CLI still pushing that branch to my normal c2 fork since I cant make another fork from a fork, maybe I mixed up something while pulling (accidentally pulling something from upstream rather than origin or something 🤷).
It's definitely fine for merging, I have double-checked and this PR does not contain any changes from upstream c2, just the reference to this one commit, so there won't be any conflicts.

AnatoleAM commented:
I can confirm you forked c2 as clicking on the branch in this PR takes you there;
![image](https://user-images.githubusercontent.com/9599344/168499942-48ce9d59-5e08-4e2a-8191-7ced480c88f9.png)

Though in theory this shouldn't be an issue as long as you pull upstream from c7.

LosFarmosCTL commented:
> I can confirm you forked c2 as clicking on the branch in this PR takes you there;

Yeah my mistake, I meant cloned, the branch is then just pushed to that repo since you can't fork the same (upstream)repo multiple times and I still need the normal c2 fork.

AnatoleAM commented:
Also regarding the "animation" tickbox; animations were deprecated after it was discovered creating a cross-platform standard for them would be too difficult. All current animated paints are actually just using the ImageURL type.

LosFarmosCTL commented:
> Also regarding the "animation" tickbox; animations were deprecated after it was discovered creating a cross-platform standard for them would be too difficult. All current animated paints are actually just using the ImageURL type.

Ah alright, yeah I had not looked into animations yet at all, just if it would be possible to implement in Qt, but only dealing with the URL type definitely makes things easier, good to know.

LosFarmosCTL committed:
fix(compiling): compiling without pch

LosFarmosCTL committed:
refactor: separate classes for 7tv badges and paints

needs 2 API calls because badges are assigned by userID while
paints need the login name

LosFarmosCTL committed:
feat(paints): add method to check if a paint needs to be rendered animated

LosFarmosCTL committed:
refactor(paints): pass the drawing rectangle to asBrush, needed for gradient calculations

LosFarmosCTL committed:
fix(paints): switch out calculation method for linear gradients to correctly translate from css angles

LosFarmosCTL committed:
feat(nametag): paint username elements in chat with the correct paint brush

LosFarmosCTL committed:
chore(formatting): run clang-format

LosFarmosCTL commented:
Figuring out how to translate the angle of a linear-gradient from css to qt's gradient system turned out to be a gigantic pain(t) in the ass, but after trying out & throwing away 10 different solutions, plus learning some more about qt's geometry classes, I now figured out how to replicate the css gradient correctly. Still needs some work on calculating the correct end point for repeating gradients, but the main problem is fixed.

Paints are now actually displayed in chat as well.

LosFarmosCTL committed:
refactor(api,paints): use QGradientStop(s) typedef

LosFarmosCTL committed:
feat(paints): handle repeating gradients correctly

LosFarmosCTL committed:
fix(paints): support hard edges in gradients

LosFarmosCTL committed:
feat(paints): calculate transparent paints based on the username color

LosFarmosCTL committed:
feat(paints): add radial-gradient paint support -WIP-

LosFarmosCTL committed:
chore(paints): remove TODO comment

LosFarmosCTL committed:
refactor(paints): move helper methods from LinearGradientPaint to parent class

LosFarmosCTL committed:
feat(api): handle transparent colors in radial gradients

LosFarmosCTL committed:
feat(paints): handle repeating radial-gradients

LosFarmosCTL committed:
chore(format): run clang-format

LosFarmosCTL committed:
feat(paints,api): load url paints (WIP)

LosFarmosCTL committed:
feat(api,paints): load drop shadows from API

LosFarmosCTL committed:
fix(paints/drop-shadows): return pointer to graphics effect object

LosFarmosCTL committed:
feat(nametag): render drop shadows on nametag

LosFarmosCTL commented:
Drop shadows are now working, still planning to look into some edge cases and rendering will definitely not be exactly the same as in browser, but imo it works well enough.

I also had a new idea for rendering animated paints which I'll try out, but no idea if it's actually going to work. If not I'll postpone that to some later day and finish up whats currently working to at least get normal paints merged in.

Excellify commented:
Animated paints are currently only used in 2 custom paints, used by 2 people, so its not a priority

LosFarmosCTL commented:
> Animated paints are currently only used in 2 custom paints, used by 2 people, so its not a priority

ye exactly anatole also told me on twitch its still kinda experimental but if my idea works they would be easy to implement, so why not.

Otherwise yeah, its not a priority and I'd happily say this PR is finished without them.

AnatoleAM commented:
> Animated paints are currently only used in 2 custom paints, used by 2 people, so its not a priority

Only 1 of them is actually animated, but both are WEBP images.

@LosFarmosCTL I updated the "Staff Shine" paint with a [lower resolution file](https://cdn.7tv.app/misc/img_paints/staff_shine2.webp).

> Otherwise yeah, its not a priority and I'd happily say this PR is finished without them.

Feel free to remove the draft status when you'd like us to check this out 👍

LosFarmosCTL commented:
> @LosFarmosCTL I updated the "Staff Shine" paint with a [lower resolution file](https://cdn.7tv.app/misc/img_paints/staff_shine2.webp).

Nice, I'll try that out later today, thanks.

> Feel free to remove the draft status when you'd like us to check this out 👍

Still have some minor things and refactoring I'd like to do, but probably won't take longer than at most a few days.

Melonify commented:
@LosFarmosCTL There's now a public image url paint, `Crazy Confetti`, just wanted to let you know just in case that changes whether or not you want to implement url paints in this patch.

LosFarmosCTL commented:
> There's now a public image url paint, Crazy Confetti, just wanted to let you know just in case that changes whether or not you want to implement url paints in this patch.

Yeah, I have just seen the reminder from anatole on twitch. I was planning to include non-animated URL paints anyways, only the animated stuff is problematic.

Already looked at the new paint, should not even need changes to what I have implemented locally right now, URL paints are just drawn above the normal username anyways, so that should all work just fine.

TroyKomodo commented:
PagMan ?

Felanbird commented:
> PagMan ?

from unrelated upstream PR comment:
![image](https://user-images.githubusercontent.com/41973452/175708906-9880a559-1755-4010-ba84-8402b0fd480f.png)

LosFarmosCTL commented:
Sorry for the lack of updates, as Felanbird reposted from upstream I have been out most of the time and basically only got home to sleep and chill a little, didn't want to actually work on anything.

I have updated the checklist in the PR description with the last things that need to be done and will have more time this weekend and next week to hopefully finish this PR.

LosFarmosCTL committed:
feat(settings): add setting to toggle off 7TV paints rendering

LosFarmosCTL committed:
feat(nametag): only render paints when the setting is enabled

LosFarmosCTL committed:
fix(nametag): only draw text element if there is something to draw

LosFarmosCTL committed:
chore(format): run clang-format

AnatoleAM commented:
Sorry I think this ci failure is my fault, my branch was affecting pull requests

LosFarmosCTL committed:
feat(paints): handle url paints

LosFarmosCTL committed:
fix(nametag): render colon separate from the nametag itself to not have the paint stretch across it

LosFarmosCTL committed:
refactor(nametag): move comment

LosFarmosCTL committed:
feat(nametag): render animated paints

LosFarmosCTL commented:
Figured out animated paints (havent separated the colon yet, I want to extract that code first to avoid duplications):

![Chatterino 2022-06-26 at 14 37 01](https://user-images.githubusercontent.com/80157503/175814602-84d29155-3706-408c-b3ff-f3554378545b.gif)

Functionality wise this PR is ready now, I just want to refactor some more things, remove code duplication, clean up, etc. Might even be finished today, but shouldn't take longer than tomorrow to have everything ready for review 😄

AnatoleAM commented:
![image](https://user-images.githubusercontent.com/9599344/175814803-838d9b59-fb6b-4aba-ab8a-0fddc8d239fc.png)

Impressive work, can't wait to try this out

AnatoleAM commented:
@Felanbird any idea why ci is still failing?

Felanbird commented:
> @Felanbird any idea why ci is still failing?

Looks like the same issue from the staging build, but I'm unsure why it's happening here.

Felanbird commented:
> but I'm unsure why it's happening here.

Actually, I have a feeling that due to the annotated tag created for the staging branch, when trying to use `--tags`, said branch is being borrowed - which causes the issue. 
![image](https://user-images.githubusercontent.com/41973452/175823421-9647b73b-8c0b-482b-a848-d4496f6c770f.png)

AnatoleAM commented:
I deleted the tag since it's no longer needed anyway, hopefully that helps

Felanbird commented:
> I deleted the tag since it's no longer needed anyway, hopefully that helps

We are probably just better off stripping `--tags` from the main build, the issue from #81 seems to be resolved without the `--tags` variable
![image](https://user-images.githubusercontent.com/41973452/175823912-5579701a-dbf2-4ec4-9da0-fbcfc876d0bf.png)

LosFarmosCTL committed:
refactor(paints,nametag): move pixmap generation into paints classes

LosFarmosCTL committed:
fix(nametag): render animated paints through the extracted getPixmap method
fixes the nametag and the colon not being rendered separately

LosFarmosCTL committed:
fix(nametag): dont draw nametag twice if its an animated paint

LosFarmosCTL committed:
chore(format): run clang-format

LosFarmosCTL committed:
fix(nametag): dont draw anything animated if the drawing rect is empty

LosFarmosCTL committed:
fuck qmake

LosFarmosCTL committed:
refactor(paints): rename variables

LosFarmosCTL committed:
refactor: make all private member variables end with an underscore

LosFarmosCTL committed:
refactor: use this-> whenever possible

LosFarmosCTL committed:
refactor(paints): directly return brush without storing in local
variable

LosFarmosCTL committed:
refactor: use const whenever possible

LosFarmosCTL committed:
chore: add newline

LosFarmosCTL committed:
chore(format): run clang-format

TroyKomodo commented:
Feel free to move it to ready and ill review.

TroyKomodo commented:
Do you mind doing a git rebase on this branch?
The steps will be like

```bash
git checkout feat/paints
git remote add c7 https://github.com/SevenTV/chatterino7
git pull c7
git rebase c7/chatterino7
git push -f
```

Should be the output
https://pastebin.com/6TiVS0uc

The reason behind this is to remove things from this pr that are not supposed to be here.
Please only do it once your branch is ready to be reviewed.

LosFarmosCTL committed:
fix(paint): fix angle calculations for angles >= 360

LosFarmosCTL committed:
fix(paint): dont apply drop shadows with a radius of 0

LosFarmosCTL committed:
chore(format): run clang-format

LosFarmosCTL commented:
> Do you mind doing a git rebase on this branch?

Sure, just didn't have the motivation to find out how exactly to get rid of that commit, rebase is something I still have to google for a lot 😄

> Feel free to move it to ready and ill review.

Fixed the last few bugs I found while testing out all paints currently available, I'll go over them once more now and if I don't find anything mark this as ready.

On a related note: Is there a particular reason for the ``80's Pool`` paint to have a drop shadow with a radius of 0? Qt apparently creates some weird render glitches in that case so I've added a check to prevent it from trying to render, but even in css it just shouldn't have any effects so seems pretty pointless to me?

![image](https://user-images.githubusercontent.com/80157503/176007860-fb1e9ada-7d97-4c4d-accf-9d97cbc7f4b2.png)

AnatoleAM commented:
> Is there a particular reason for the 80's Pool paint to have a drop shadow with a radius of 0

Definitely a mistake, I'll change that.

LosFarmosCTL committed:
fix(paints/nametag): scale drop shadow

LosFarmosCTL committed:
refactor(paints): mark getPixmap method as const

LosFarmosCTL commented:
LGTM:

![image](https://user-images.githubusercontent.com/80157503/176009942-468cf53f-13fe-45ee-bf9c-e9e14913309f.png)

Festive Gold doesn't really look very nice sadly, but from what I can tell thats on Qt's drop shadow rendering, which just isn't as smooth as the browser one.

TroyKomodo commented:
@LosFarmosCTL Seems good, whisper doesnt work yet tho.
![image](https://user-images.githubusercontent.com/49777269/176100209-1fb10d98-eaf5-4ed0-91d4-7f7b992d7ae3.png)
And also sending whispers doesnt work might not be part of your pr since well nothing works on sending.
![image](https://user-images.githubusercontent.com/49777269/176100271-f3529cd2-473b-4fbf-9888-fce25a259fb3.png)
Very nice work!
Something i noticed
```diff
diff --git a/src/providers/seventv/paints/LinearGradientPaint.cpp b/src/providers/seventv/paints/LinearGradientPaint.cpp
index 77dc774e..49789526 100644
--- a/src/providers/seventv/paints/LinearGradientPaint.cpp
+++ b/src/providers/seventv/paints/LinearGradientPaint.cpp
@@ -59,8 +59,8 @@ QBrush LinearGradientPaint::asBrush(const QColor userColor,
 
     QPointF gradientStart;
     QPointF gradientEnd;
-    gradientAxis.intersects(colorStartAxis, &gradientStart);
-    gradientAxis.intersects(colorStopAxis, &gradientEnd);
+    gradientAxis.intersect(colorStartAxis, &gradientStart);
+    gradientAxis.intersect(colorStopAxis, &gradientEnd);
 
     if (this->repeat_)
     {
```

For whatever reason my version of qt had a different header files, not sure if this is documented could u provide some clarity on which is newest.

~~And also under building for linux, the package is not `qt5-image-formats-plugin` for ubuntu its `qt5-image-formats-plugins`
could u also update that in this pr.~~ Thanks @Felanbird

TroyKomodo commented:
Oh and 1 more small side note, not a big deal
![image](https://user-images.githubusercontent.com/49777269/176100874-63ad3ee1-d586-4050-ad7b-ca13d862c96f.png)
![image](https://user-images.githubusercontent.com/49777269/176100918-48eb1454-fa5c-44af-b02b-a5dbb4328291.png)
when we don't have paints enabled the colon is colored too.
Is there a reasoning behind not coloring it such as readability or something.

TroyKomodo commented:
Other than that really good work man thank you a lot for your contributions please tag us in discord when u read this and we will set u up with the contributors badge and role <3 We really appreciate the effort on this!

Felanbird commented:
> Is there a reasoning behind not coloring it such as readability or something.

Looks to be to match webchat behavior | also efb4aec 

![image](https://user-images.githubusercontent.com/41973452/176101471-a3b705b2-96f1-422b-9755-f0285322484e.png)

Felanbird commented:
> And also under building for linux, the package is not `qt5-image-formats-plugin` for ubuntu its `qt5-image-formats-plugins` could u also update that in this pr.

#98

LosFarmosCTL committed:
fix(nametag): render paints for whispers as well

LosFarmosCTL commented:
> For whatever reason my version of qt had a different header files, not sure if this is documented could u provide some clarity on which is newest.

Seems like it was called ``intersect`` up until Qt 5.14, actually not sure how to handle this since Qt 5.12 is still supposed to be supported. @Felanbird you know any examples from upstream for situations like that?

https://doc.qt.io/archives/qt-5.12/qlinef.html#intersect
https://doc.qt.io/qt-5/qlinef.html#intersects


> Is there a reasoning behind not coloring it such as readability or something.

Yeah it's basically just me looking at how it's handled in webchat and matching that, since the paints extending over the colon in chatterino would change their appearance compared to webchat. IMO it only makes sense for nametags that actually do have a paint rendered on them, since the colored colon might conflict with the paints design (thats why I didn't change anything about how nametags without paints are drawn, didn't want to change c2's general design there)


> whisper doesnt work yet tho.

Fixed what I could, for outgoing whispers + your own name when receiving there is no link information included, which is what we need to determine if whatever is drawn right now is actually a nametag element. Might make sense to add one, but that would be an upstream PR (if those are added, paints should be rendered correctly without further changes)


> Other than that really good work man thank you a lot for your contributions please tag us in discord when u read this and we will set u up with the contributors badge and role <3 We really appreciate the effort on this!

Thanks a lot for that feedback :) This PR was actually really fun to work on, learned a ton of stuff while creating it as well.

Felanbird commented:
> Seems like it was called intersect up until Qt 5.14, actually not sure how to handle this since Qt 5.12 is still supposed to be supported. @Felanbird you know any examples from upstream for situations like that?

No idea, but it's likely we can sort of ignore 5.12 as it's already been dropped here due to to issues with the animated pfp build.
5.12 having issues with 2/2 of the Major PRs - as well as I believe Qt6 has been discussed more recently, so by the time `chatterino7 -> chatterino2` happens we might just not have to deal with it. 😁
